### ISSUE TITLE
Linux Shortcuts don't work with non-latin / international keyboard layouts

### ISSUE THREAD

**Issue by leonovk** [zed-industries/zed#14053]
### Description

If you switch the keyboard layout to another language, the hotkeys stop working.

For example ctrl + s or ctrl + c.

### Environment

Zed on Linux

**Comment by notpeter** [comment]
Can you provide some additional details:
1. Which desktop environment are you using?
2. Which tool are you using to switch keyboard layouts (setxkbmap, loadkeys, localectl, /etc/default/keyboard, etc)?
3. Which language(s) or keyboard layout(s) are causing you trouble?
4. What's the language/locale of your keyboard (QWERTY, QWERTZ, AZERTY, ÄWERTY, etc)

**Comment by leonovk** [comment]
1. GNOME 46 Wayland
2. To switch the keyboard layout, I use the default hotkey in the Gnome environment "super + spacebar". The language switches, but the hotkeys inside zed stop working.
3. Hotkeys do not work when the Russian keyboard is on
4. My keyboard has 2 layouts: Russian and English (QWERTY/ЙЦУКЕН)

**Comment by AndyDroid10X** [comment]
KDE 6.1.2 Wayland
The same issue, the same layouts

**Comment by yournumberone** [comment]
same issue. ubuntu 24.04 gnome 46, qwerty/йцукен (eng/rus)

**Comment by ruskiiamov** [comment]
GNOME 42.9, Wayland
Ubuntu 22.04.4
2 layouts: Russian and English (QWERTY/ЙЦУКЕН)
The same issue

**Comment by Beryesa** [comment]
Not using the Russian layout but, using keycodes for stuff like the \` key instead of the actual character would help here as well
i.e. if we can press the "key under esc" regardless of the layout.
Because on mine the \` character is at AltGr + "," so doing Ctrl+AltGr+, is quite inconvenient :sweat_smile:

**Comment by notpeter** [comment]
> MacOS has the same issue with shortcuts that use `Opt` key (for example `Opt+T` for a task run). Tested on MacBook Air M1, Sonoma 14.5

Please do not hijack this one with off-topic comments, this is a linux specific keyboard layout issue. Please open a new issue if you're having trouble.

FYI your issue is likely because `Opt+T` is invalid. 1. Zed use lowercase `alt` on all platforms rather than the MacOS specific `Option` or `Opt`. 2. Zed use `-` not `+`.

Any of these should work: `alt-t`, `alt-shift-t`, `alt-T` -- if not open a new issue.

**Comment by brookback** [comment]
I have the same/similar issue, but on macOS, using Zed v0.149.0. Let me know if this goes into this existing issue or if I should file a new one.

- For me, it's **switching window focus stopped working in Zed** (it works elsewhere in macOS). It's <kbd>cmd+`</kbd>.
- I can still use `cmd+s, cmd+c` etc, however. So it's not as severe as OP's issue.
- I'm using Swedish as input language.
- If I switch back to English (US) input language, I cannot repro.

**macOS keyboard settings**
<img width="457" alt="Screenshot 2024-08-15 at 08 34 03" src="https://github.com/user-attachments/assets/bf1b1274-e467-421a-84fd-822c08e49ca9">

1. **Which desktop environment are you using?** macOS
2. **Which tool are you using to switch keyboard layouts (setxkbmap, loadkeys, localectl, /etc/default/keyboard, etc)?** None.
3. **Which language(s) or keyboard layout(s) are causing you trouble?** Swedish.
4. **What's the language/locale of your keyboard (QWERTY, QWERTZ, AZERTY, ÄWERTY, etc)**. QWERTY.

**Zed specs**
Zed: v0.149.0 (Zed Preview)
OS: macOS 14.4.0
Memory: 32 GiB
Architecture: aarch64

**Comment by notpeter** [comment]
@brookback Please file a new issue as your issue on MacOS is likely distinct. Please include full reproduction steps if possible. Thanks.

**Comment by D3st0ny** [comment]
same issue. nixos-unstable hyprland 0.42.0, qwerty/йцукен (eng/rus)

**Comment by notpeter** [comment]
There is an active discussion going on here:
- https://github.com/zed-industries/zed/discussions/17031

We'd love to gather input from folks who are using international keyboards with Zed.

**Comment by ConradIrwin** [comment]
v0.162.x fixes most problems for macOS, so now Linux...

The goal here is to make the gpui platform implementation consistent with the macOS one, so:

* For layouts which generate a different key when the platform modifier is down, we should use that as the basis of the keybindings. (so, ctrl-я becomes ctrl-q)
* For layouts with dead-keys, we should generate a shortcut press for the dead key (so that ctrl-^ works, and even ^ to go to start of line in vim mode); we also need to generate shortcuts before IME if an IME is not in progress (so vim mode works on Japanese keyboards).
* We also want to copy the changes for Extended Latin keyboards. For macOS, we copy the key equivalents from the OS; for Linux I'm not sure we can copy equivalents from anywhere (is there already a convention about what to do about ctrl-[ when [ requires alt to type?). We could possibly copy VSCode's logic if we want to avoid creating our own.

I took a bit of a look, and found that the first problem is that for non-Latin layouts that should have a different character when the platform modifier is pressed, we don't correctly read that even though `xev` does.

If anyone has done key handling on Linux before, I'd love to pair with you: https://calendly.com/conradirwin/pairing.

**Comment by VyacheslavKuchumov** [comment]
yeaaaahhh i have the same issue qwerty/йцукен (eng(US)/rus)

**Comment by ConradIrwin** [comment]
@VyacheslavKuchumov are you on macOS or Linux? If Linux, are you on Wayland or X11, and which IME system do you use?

**Comment by ConradIrwin** [comment]
@ExposedCat I'd like this to work by default - so we don't need mappings in most cases.  How other apps on Linux work; but it's not totally clear to me (and probably not as standardized as it is on macOS).

For example I know if I use a Cyrillic keyboard, then shortcuts like ctrl-shift-n for new window in Gnome terminal still work. Is that because they are bound to the key code, or is it because they are bound to a layer on your keyboard that generates ASCII still?

**Comment by ExposedCat** [comment]
@ConradIrwin can [this](https://github.com/torvalds/linux/blob/master/include/uapi/linux/input-event-codes.h) be used?

**Comment by RasmusHogslatt** [comment]
Has this issue been resolved? I am using a Macbook Pro M4 with a Swedish layout. Shortcuts like toggling line comments do not work with swedish input.

**Comment by ConradIrwin** [comment]
@RasmusHogslatt on a Swedish keyboard toggle comments is now cmd-backtick

<img width="590" alt="Screenshot 2024-12-06 at 09 52 42" src="https://github.com/user-attachments/assets/41a27614-1fc3-4238-b8e0-bd86bc3c91fc">

**Comment by RasmusHogslatt** [comment]
@ConradIrwin I got back and tried this. Your comment was correct, and it reflects my setup, but for whatever reason, instead of toggling comments it opens the "Help" window for zed in the top menubar. Is that expected behaviour?

**Comment by ConradIrwin** [comment]
@RasmusHogslatt that is *not* expected; part of the reason to move these shortcuts around is to avoid the builtin macOS shortcuts.

Are you also using the Swedish locale? What's the default help menu shortcut?

**Comment by ConradIrwin** [comment]
Should we adjust the macOS Swedish layout equivalents to be less intrusive by default?

```
        "com.apple.keylayout.Swedish" => &[
   // Copied from macOS
            // Move the window switching shortcuts to match the default Swedish locale
            ('`', '<'),
            ('~', '>'),
            ('<', ';'),
            ('>', ':'),
            (':', 'Å'),
            (';', 'å'),
            // Move these untypable "pairs" to typeable pairs (so indent/outdent feel similar)
            ('{', 'Ö'),
            ('}', 'Ä'),
            ('[', 'ö'),
            (']', 'ä'),
            // keep cmd-shift-3 in the same place as it's used for macOS screenshots by default
            ('$', '€'),
  
   // Deviate from macOS (and current Zed)     
            // There are three remaining ASCII characters that require option to type on a Swedish keyboard, 
            // assign these to the three remaining characters that don't require option to type.
            ('@', '¨'),
            ('\\', '´'),
            ('|', '`'),

            // remove all these...
            // ('"', '^'),
            // ('&', '/'),
            // ('(', ')'),
            // (')', '='),
            // ('*', '('),
            // ('^', '&'),
            // ('/', '´'),
            // ('=', '`'),
            // ('\'', '¨'),
        ],
```

What is the default help shortcut in the Swedish locale? I also want to make sure we're not going to crash into it.


See also https://github.com/zed-industries/zed/issues/20425#issuecomment-2533543683 for similar woes with German and a proposed new default.

**Comment by bnn1** [comment]
same issue with en/ru kb layout.

endeavouros arch distro, x11 desktop (awesome)

**Comment by Akiyamka** [comment]
I have same issue, shortcuts like `ctrl+c` and `ctrl+v` etc. Works only when active system layout is QWERTY, any other non latin layout breaks shortcuts.  
Zed version: `0.166.1`  
OS: Fedora Linux 41 (Wayland)   
This issue not reproduced in other editors, vscode for example.   
Would like have an option that switches all shortcuts to keycode bind instead of char

**Comment by neodraw** [comment]
There was a similar problem for Cyrillic keyboard on Ubuntu. I share the solution

1) Ctrl+Shift+P --> (execute command) debug: open key context view

2) Press all the necessary key combinations and save their designations
For example, Last Keystroke: Typed: "ctrl-cyrillic_es" ..... and others

3) Menu - Open Key BIndings - add combinations
(For example for cyrillic keyboard)
{
"context": "Editor",
"bindings": {
"ctrl-cyrillic_es":"editor::Copy",
"ctrl-cyrillic_em":"editor::Paste",
"ctrl-cyrillic_che": "editor::Cut",
"ctrl-cyrillic_ef":"editor::SelectAll",
"ctrl-cyrillic_ya": "editor::Undo",
"ctrl-cyrillic_en": "editor::Redo", 
"ctrl-cyrillic_a": "buffer_search::Deploy"
 }
 }
4) worked

**Comment by Akiyamka** [comment]
@neodraw - amazing, thank you

**Comment by Randommist** [comment]
I wrote a script that translates shortcuts into Cyrillic, so you can use or copy the already translated file `localized_keymap.json` here is the link to the script https://github.com/Randommist/zed-cyrillic

**Comment by loonlylokly** [comment]
A simple space still doesn't work(

**Comment by ziimakc** [comment]
this is super annoying, maybe there is a fallback like option for zed to avoid workarounds?

**Comment by viktorandreevichmorozov** [comment]
Same issue on linux Ubuntu 20.04.6 , ukrainian language, Zed 0.186.8.

**Comment by cobisimo** [comment]
Here are some shortcuts I mostly use for Serbian Cyrillic based on @neodraw 's solution:

```json
  {
    "context": "(Workspace && keyboard_layout == Serbian) > Pane > Editor",
    "bindings": {
      "super-cyrillic_es": "workspace::SaveAll",
      "ctrl-cyrillic_es": "workspace::Save",
      "ctrl-cyrillic_zhe": "editor::Undo",
      "ctrl-cyrillic_ze": "editor::Redo",
      "ctrl-shift-cyrillic_zhe": "editor::Redo",
      "ctrl-cyrillic_dzhe": "editor::Cut",
      "ctrl-cyrillic_tse": "editor::Copy",
      "ctrl-cyrillic_ve": "editor::Paste",
      "ctrl-cyrillic_a": "editor::SelectAll",
      "ctrl-cyrillic_ef": "buffer_search::Deploy",
      "ctrl--": "editor::ToggleComments",
      "ctrl-cyrillic_de": ["editor::SelectNext", { "replace_newest": false }],
      "ctrl-shift-cyrillic_pe": "command_palette::Toggle",
      "ctrl-shift-cyrillic_i": "editor::Format"
    }
  }
```

**Comment by afiskon** [comment]
I'm using Ubuntu 24.04 LTS with en/ru keyboard layouts and this is very annoying. Firefox, Gnome Terminal etc handle all the hotkeys in an alternative layout properly.

**Comment by JohnBat26** [comment]
https://github.com/Randommist/zed-cyrillic/blob/main/localized_keymap.json

**Comment by ConradIrwin** [comment]
If anyone is following this thread and able to test Zed v0.196.4 (the latest preview), it should now be the case that (most) shortcuts do not need to be localized for non-Latin keyboards.

There is still more work to do, as currently we only fall-back when the key is a non ASCII character, but would be good to know if there are still unknown prbolems.

**Comment by afiskon** [comment]
Hi @ConradIrwin,

I tried v0.196.4 and can confirm that the hotkeys work in an alternative layout properly.

Many thanks to everyone involved!

**Comment by cobisimo** [comment]
I can also confirm. All shortcuts I tried work fine in Serbian Cyrillic. Thanks @ConradIrwin !

**Comment by rakhmaevao** [comment]
My hotkeys also work on both keyboard layouts. Thanks.

**Comment by evtn** [comment]
@ConradIrwin, it's great that Zed has finally achieved this point, but the "non-ASCII fallback" is for sure only a half of solution. 
There are many symbol keys on non-Latin keyboard that are placed vastly different to, for example, English (US). 

For example, in US English layout, pressing `Shift+4` yields `$`, while in Russian the same combination yields `;` (semicolon). 

And, finally, can we get a settings flag to only rely on scancodes in shortcuts, mapping to en-us qwerty? That way, it could be easily enabled and will mostly solve a lot of problems for users of several layouts.

**Comment by ConradIrwin** [comment]
I think so, and now I understand things more, I think the next step is to read the current keyboard layout when loading the keyboard and (when `use_key_equivalents` is set) relabel the shortcuts to match the same key by the X11/Wayland key-code that they would have had. (e.g. ctrl-$ would be loaded as ctrl-; so you can type them using the same key).

**Comment by douglasrizzo** [comment]
In my case, I use the US international layout. Opening a terminal is mapped to "Ctrl + `", but I have to Ctrl + ` twice to open the terminal, and the first time it actually inserts the ` anywhere the editor is focused.

Zed: v0.196.6 (Zed) 
OS: Linux X11 arch unknown
Memory: 15.5 GiB
Architecture: x86_64
GPU: NVIDIA GeForce GTX 1650 || NVIDIA || 575.64.05

**Comment by thraizz** [comment]
I use a custom macOS keyboard layout and with this, I still cannot open the Command Palette with Cmd+Shift+P or do other modifications (CMD+P itself works). This works fine in VSCode, does not work in Zed. If I switch to the Apple default English layout, it works.

**Comment by ConradIrwin** [comment]
@thraizz What version of Zed are you using, and what is your keyboard layout?

**Comment by thraizz** [comment]
> [@thraizz](https://github.com/thraizz) What version of Zed are you using, and what is your keyboard layout?

I am on Zed 0.202.7. But I reinstalled my keyboard layout and rebooted and everything works flawless now. I had this originally described here: https://github.com/zed-industries/zed/issues/24643#issuecomment-2720776424

Thank you for fixing and unblocking Zed for me <3

**Comment by kraftwer1** [comment]
(For the github-actions bot) the issue is still relevant because it‘s still not possible to bind to umlauts and possibly other international special characters either. It would be much appreciated if this issue could be fixed as it would boost my daily workflow.

**Comment by miguelraz** [comment]
Sorry for the noise, the `stalebot` upstream went haywire.

Please disregard it.

**Comment by AndriiShkrobutGL** [comment]
I'm not sure if it's related, but I have two layouts: English (UK) and Ukrainian.
When I'm on English layout and I press `ctrl + .` for code actions, it inserts `e` into code, and I can't backspace it; I can only `esc` to remove it. However, the code actions keymap works fine when I switch to the Ukrainian layout.

Environment:
`Ubuntu 22.04.5 LTS` x86_64
 `Zed 0.213.3` 


UPD: when I try to remap or search for a keymap for code actions in the keymap editor, it just refuses to record the `.` key on English layout, it reads `Ctrl-` only


UPD 2 & Solution: it's related to https://github.com/zed-industries/zed/discussions/27010 and the issue is that `ctrl-.` is a shortcut for emoji. Here's how to turn it off: https://unix.stackexchange.com/a/694244

**Comment by seet61** [comment]
Zed 0.232.2 still a problem

**Comment by joaoantoniocardoso** [comment]
Zed `1.3.5`

still a problem

update: `1.12.0` and still a problem

**Comment by mTvare6** [comment]
Can't reproduce on master on `x11/dwm` or `wayland/hyprland` on Russian.

**Comment by jerrygreen** [comment]
It was partially fixed in some update. Hotkeys like `ctrl+c` or `ctrl+s` indeed work.

But others like `ctrl+/` still don't work. An example of a problematic hotkey:

```json
"ctrl-/": ["editor::ToggleComments", { "advance_downwards": false }],
```

Not sure if it's the last one that is problematic key, but others like `ctrl+{` `ctrl+'` `ctrl+\"` `ctrl+.` – all seem to work, apparently.

<kbd>/</kbd> seems to be one of the last problematic keys. If not latest.

**Comment by mTvare6** [comment]
> It was partially fixed in some update. Hotkeys like `ctrl+c` or `ctrl+s` indeed work.
> 
> But others like `ctrl+/` still don't work. An example of a problematic hotkey:
> 
> "ctrl-/": ["editor::ToggleComments", { "advance_downwards": false }],
> Not sure if it's the last one that is problematic key, but others like `ctrl+{` `ctrl+'` `ctrl+\"` `ctrl+.` – all seem to work, apparently.
> 
> / seems to be one of the last problematic keys. If not latest.

The issue seems to be that the Russian map produces `.` on pressing `/` and the current mapping system doesn't have to produce the `/` as it has a valid ascii `.` symbol. I'll look into a more general solution than the current one which seems to not be very robust.

**Comment by Ciantic** [comment]
I can bind Ctrl+ö (it appears in the Zed as Ctrl+;).
But I cannot bind Ctrl-Shift-ö, it doesn't work at all...

VSCode in linux understands the letters correctly, it shows "Ctrl+ö" and "Ctrl+shift+ö", I think it would be much easier if it could be figured out exact same way, trying to translate the keyboard to US first makes things kind of messy.

**Comment by Nikita0x** [comment]
Incredibly, it was working for me for `ru` and `uk` layouts - but I reinstalled linux - and now I have this problem again and I don't remember how I fixed it or which workaround I used....

**Comment by mTvare6** [comment]
@jerrygreen @Nikita0x my Russian layout makes `Shift+|` into `/`, and `Ctrl+Shift+|` toggles comments. The same behaviour is replicable across VSCode. Do you still think the current behaviour is broken?

**Comment by Nikita0x** [comment]
> [@jerrygreen](https://github.com/jerrygreen) [@Nikita0x](https://github.com/Nikita0x) my Russian layout makes `Shift+|` into `/`, and `Ctrl+Shift+|` toggles comments. The same behaviour is replicable across VSCode. Do you still think the current behaviour is broken?

No, I don't think it is broken. 
Actually I found the issue (or, to be frank, ChatGPT).

I am on Linux Mint 21.3 - and by default it installs `ibus` - an input framework for complex layouts (chinese, japanese etc) - and it overtakes input events so that `dev:  open key context view` receives nothing. 

Since I don't use chinese/japanese layouts - I removed `ibus` completely from the system - and now it works flawlessly for `ru`, `uk` and `en`.
